### PR TITLE
network: about xxx_configured flags

### PR DESCRIPTION
**Pull request by yuwata**
Follow-ups for #16379.

### CODE REVIEW

**Comment by yuwata** [comment]
@ssahani I'd like to add test for `AssignAcquiredDelegatedPrefixAddress=`. How should I configure dnsmasq for that?

**Comment by ssahani** [comment]
We can't . This requires a ISP setup . I think @mamarley   can help us.

**Comment by mamarley** [comment]
Yeah, I put together a test jig for a DHCPv6-PD setup.  If you tell me what to look for here I can test this.

**Comment by keszybz** [comment]
CI fails:
```
======================================================================
FAIL: test_coldplug_dhcp_ip6 (__main__.DnsmasqClientTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.Jh1tAM/build.6CR/systemd/test/networkd-test.py", line 463, in test_coldplug_dhcp_ip6
    self.do_test(coldplug=True, ipv6=True)
  File "/tmp/autopkgtest.Jh1tAM/build.6CR/systemd/test/networkd-test.py", line 401, in do_test
    self.assertRegex(out, (r'{}\s+ether\s+routable\s+configured'.format(self.iface)).encode())
AssertionError: Regex didn't match: b'test_eth42\\s+ether\\s+routable\\s+configured' not found in b'IDX LINK         TYPE     OPERATIONAL SETUP      \n  1 lo           loopback carrier     unmanaged  \n  2 ens2         ether    routable    unmanaged  \n 19 router_eth42 ether    routable    unmanaged  \n 20 test_eth42   ether    routable    configuring\n\n4 links listed.\n'

======================================================================
FAIL: test_hotplug_dhcp_ip6 (__main__.DnsmasqClientTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.Jh1tAM/build.6CR/systemd/test/networkd-test.py", line 470, in test_hotplug_dhcp_ip6
    self.do_test(coldplug=False, ipv6=True)
  File "/tmp/autopkgtest.Jh1tAM/build.6CR/systemd/test/networkd-test.py", line 401, in do_test
    self.assertRegex(out, (r'{}\s+ether\s+routable\s+configured'.format(self.iface)).encode())
AssertionError: Regex didn't match: b'test_eth42\\s+ether\\s+routable\\s+configured' not found in b'IDX LINK         TYPE     OPERATIONAL SETUP      \n  1 lo           loopback carrier     unmanaged  \n  2 ens2         ether    routable    unmanaged  \n 27 router_eth42 ether    routable    unmanaged  \n 28 test_eth42   ether    routable    configuring\n\n4 links listed.\n'

======================================================================
FAIL: test_resolved_etc_hosts (__main__.DnsmasqClientTest)
resolved queries to /etc/hosts
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.Jh1tAM/build.6CR/systemd/test/networkd-test.py", line 700, in test_resolved_etc_hosts
    self.do_test(coldplug=None, ipv6=True)
  File "/tmp/autopkgtest.Jh1tAM/build.6CR/systemd/test/networkd-test.py", line 401, in do_test
    self.assertRegex(out, (r'{}\s+ether\s+routable\s+configured'.format(self.iface)).encode())
AssertionError: Regex didn't match: b'test_eth42\\s+ether\\s+routable\\s+configured' not found in b'IDX LINK         TYPE     OPERATIONAL SETUP      \n  1 lo           loopback carrier     unmanaged  \n  2 ens2         ether    routable    unmanaged  \n 33 router_eth42 ether    routable    unmanaged  \n 34 test_eth42   ether    routable    configuring\n\n4 links listed.\n'

----------------------------------------------------------------------
Ran 36 tests in 156.359s

FAILED (failures=3, skipped=2)
autopkgtest [09:27:07]: test networkd-test.py: -----------------------]
networkd-test.py     FAIL non-zero exit status 1
```

**Comment by yuwata** [comment]
Updated. I hope all tests will pass now. @mamarley Could you test this PR? Thank you for your contribution.

**Comment by mamarley** [comment]
It looks like the CentOS CI stuff is still failing due to an abort in networkd.  I will hold off testing until the CI passes.

**Comment by yuwata** [comment]
Ouch. Updated again.

**Comment by keszybz** [comment]
centos ci (arch in kvm with sanitizers):
```
22:29:34 ======================================================================
22:29:34 ERROR: test_wireguard (__main__.NetworkdNetDevTests)
22:29:34 ----------------------------------------------------------------------
22:29:34 Traceback (most recent call last):
22:29:34   File "test/test-network/systemd-networkd-tests.py", line 1163, in test_wireguard
22:29:34     self.wait_online(['wg99:carrier', 'wg98:routable', 'wg97:carrier'])
22:29:34   File "test/test-network/systemd-networkd-tests.py", line 545, in wait_online
22:29:34     check_output(*args, env=env)
22:29:34   File "test/test-network/systemd-networkd-tests.py", line 49, in check_output
22:29:34     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
22:29:34   File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
22:29:34     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
22:29:34   File "/usr/lib/python3.8/subprocess.py", line 512, in run
22:29:34     raise CalledProcessError(retcode, process.args,
22:29:34 subprocess.CalledProcessError: Command '['/systemd-meson-build/systemd-networkd-wait-online', '--timeout=20s', '--interface=wg99:carrier', '--interface=wg98:routable', '--interface=wg97:carrier']' returned non-zero exit status 1.
```

**Comment by yuwata** [comment]
> centos ci (arch in kvm with sanitizers):
> 
> ```
> 22:29:34 ======================================================================
> 22:29:34 ERROR: test_wireguard (__main__.NetworkdNetDevTests)
> 22:29:34 ----------------------------------------------------------------------
> 22:29:34 Traceback (most recent call last):
> 22:29:34   File "test/test-network/systemd-networkd-tests.py", line 1163, in test_wireguard
> 22:29:34     self.wait_online(['wg99:carrier', 'wg98:routable', 'wg97:carrier'])
> 22:29:34   File "test/test-network/systemd-networkd-tests.py", line 545, in wait_online
> 22:29:34     check_output(*args, env=env)
> 22:29:34   File "test/test-network/systemd-networkd-tests.py", line 49, in check_output
> 22:29:34     return subprocess.check_output(command, universal_newlines=True, **kwargs).rstrip()
> 22:29:34   File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
> 22:29:34     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
> 22:29:34   File "/usr/lib/python3.8/subprocess.py", line 512, in run
> 22:29:34     raise CalledProcessError(retcode, process.args,
> 22:29:34 subprocess.CalledProcessError: Command '['/systemd-meson-build/systemd-networkd-wait-online', '--timeout=20s', '--interface=wg99:carrier', '--interface=wg98:routable', '--interface=wg97:carrier']' returned non-zero exit status 1.
> ```

It seems to be unrelated.
```
wg97: Could not bring up interface: Address already in use
```

**Comment by yuwata** [comment]
Force-pushed to restart CIs.

**Comment by mamarley** [comment]
Thanks, this seems to work fine in my test environment.

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -1000,8 +998,6 @@ static int route_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link) {
 
         assert(link);
         assert(link->route_messages > 0);
-        assert(IN_SET(link->state, LINK_STATE_CONFIGURING,
-                      LINK_STATE_FAILED, LINK_STATE_LINGER));
```
Comment (ssahani) [inline src/network/networkd-link.c]: Why remove this ? Is the race still there

File: src/network/networkd-dhcp6.c [hunk 2]
```diff
@@ -1058,11 +1110,13 @@ static int dhcp6_assign_delegated_prefix(Link *link,
         address->cinfo.ifa_prefered = lifetime_preferred;
         address->cinfo.ifa_valid = lifetime_valid;
 
-        link_set_state(link, LINK_STATE_CONFIGURING);
```
Comment (ssahani) [inline src/network/networkd-dhcp6.c:1061]: dropped the setting state ?

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -948,8 +948,6 @@ static int nexthop_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link)
 
         assert(link);
         assert(link->nexthop_messages > 0);
-        assert(IN_SET(link->state, LINK_STATE_CONFIGURING,
-                      LINK_STATE_FAILED, LINK_STATE_LINGER));
 
         link->nexthop_messages--;
```
Comment (ssahani) [inline src/network/networkd-link.c]: same here

File: src/network/networkd-link.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-link.c]: `link_request_set_routes()` is called when dynamic addresses are configured. In that time, the link state may be already in configured state.

File: src/network/networkd-dhcp6.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-dhcp6.c:1061]: We never set link state like this in other places. Please see networkd-dhcp4.c or networkd-ndisc.c for example.

File: src/network/networkd-link.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-link.c]: This change was dropped. This is not necessary to remove this assertion. Thanks.

File: src/network/networkd-link.c (hunk 1 above)
Comment (ssahani) [reply src/network/networkd-link.c]: This means the race is still there. We count the ```route_messages```  and when this becomes zero we make it configure.  What is condition why it getting configured ?

File: src/network/networkd-dhcp6.c (hunk 2 above)
Comment (ssahani) [reply src/network/networkd-dhcp6.c:1061]: This is the race we trying to fix. By defining different type of messages and a boolean value to set it true and check in ```link_check_ready```. If we decide to remove here the Just  doing ++ ```dhcp6_assign_delegated_prefix``` prefix messages  and doing -- solve the issue.

File: src/network/networkd-link.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-link.c]: Sorry, I noticed that `link_request_set_routes()` calls `link_set_state(link, LINK_STATE_CONFIGURING)`. So it is not necessary to remove this assertion. Now this change is dropped. Thanks.

File: src/network/networkd-dhcp6.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-dhcp6.c:1061]: After testing more, I noticed (but still I do not know why) a DHCPv6 address assigned in test-network.py (not systemd-networkd-tests.py) through dnsmasq stick to tentative state... So, setting `LINK_STATE_CONFIGURED` here or at one more place cause the link state stick to configuring state and never become to configured state.

**Comment by yuwata** [comment]
Finally (I hope) I found the root of the strange behavior in test-network.py. Now I fixed the tests and added `link_set_state()` again. PTAL.

**Comment by ssahani** [comment]
Thanks @yuwata .This looks much better. I hope this will fix the race. @mamarley ?

**Comment by yuwata** [comment]
@ssahani Thank you for your comments.

@mamarley I've changed this PR a lot after once you tested it. If possible, could you test this again?

**Comment by mamarley** [comment]
It seems to work fine in my test environment.  I have let it renew the lease several times and restarted it manually several times without any bad behavior.  Thanks!

**Comment by yuwata** [comment]
@mamarley Thank you for your cooperation!

**Comment by yuwata** [comment]
@keszybz Could you take a final look?

**Comment by yuwata** [comment]
Hmm, the last commit reveal another race...

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -1085,71 +1085,110 @@ void link_check_ready(Link *link) {
 
         assert(link);
 
-        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER)) {
+                log_link_debug(link, "%s(): link is in failed or liner state.", __func__);
```
Comment (keszybz) [inline src/network/networkd-link.c]: "linger"

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -1168,6 +1168,11 @@ void link_check_ready(Link *link) {
                         return;
                 }
 
+                if ((link_dhcp4_enabled(link) || link_dhcp6_enabled(link)) && set_isempty(link->addresses)) {
+                        log_link_debug(link, "%s(): DHCP4 or DHCP6 is enabled but no address is assigned yet.", __func__);
```
Comment (keszybz) [inline src/network/networkd-link.c:1163]: Sounds reasonable.

File: src/network/networkd-link.c (hunk 4 above)
Comment (yuwata) [reply src/network/networkd-link.c]: Ouch. Fixed.

**Comment by mamarley** [comment]
I tested the PR again and everything is still working fine.

**Comment by keszybz** [comment]
bionic-i386 timed out, looks unrelated.

**Comment by yuwata** [comment]
@mamarley and @keszybz Thanks!

**Pull request merged**